[lldb] Enable chaining multiple scripted frame providers per thread

medismailben opened the pull request:
This patch allows threads to have multiple SyntheticFrameProviderSP instances that chain together sequentially. Each provider receives the output of the previous provider as input, creating a transformation pipeline.

It changes `Thread::m_frame_provider_sp` to a vector, adds provider parameter to SyntheticStackFrameList to avoid calling back into `Thread::GetFrameProvider()` during frame fetching, updated `LoadScriptedFrameProvider()` to chain providers by wrapping each previous provider's output in a `SyntheticStackFrameList` for the next provider and finally, loads ALL matching providers in priority order instead of just the first one.

The chaining works as follows:
```
  Real Unwinder Frames
      ↓
  Provider 1 (priority 10) → adds/transforms frames
      ↓
  Provider 2 (priority 20) → transforms Provider 1's output
      ↓
  Provider 3 (priority 30) → transforms Provider 2's output
      ↓
  Final frame list shown to user
  ```

This patch also adds a test for this (test_chained_frame_providers) to verify that 3 providers chain correctly: `AddFooFrameProvider`, `AddBarFrameProvider`, `AddBazFrameProvider`.

bulbazord commented on lldb/source/Target/Thread.cpp:
```diff
@@ -1506,29 +1506,39 @@ llvm::Error Thread::LoadScriptedFrameProvider(
+    // Subsequent providers get the previous provider's OUTPUT
+    // We create a SyntheticStackFrameList that wraps the previous provider
+    SyntheticFrameProviderSP prev_provider = m_frame_providers.back();
+    StackFrameListSP prev_input = prev_provider->GetInputFrames();
```
Something here feels strange. The comment saying you're getting the previous provider's output but then you call this variable prev_input. It _sounds_ like `prev_input` represents the frame list that the previous provider received? Or is it supposed to be the input *from* the previous provider?

I think I understand what's supposed to happen but I'm having trouble following the implementation logic here.

bulbazord reviewed (commented):
I think I follow what this PR is supposed to do, but I had some trouble following the implementation. I added some comments inline to explain where I was confused.

As an overall suggestion, I think you should add a test case for providers with the same priority level. I'm not sure what the intended order is in that case. If any order is possible, you should write a test that checks all scenarios.

JDevlieghere commented on lldb/include/lldb/Target/Thread.h:
```diff
@@ -1409,8 +1409,8 @@ class Thread : public std::enable_shared_from_this<Thread>,
-  /// The Scripted Frame Provider, if any.
-  lldb::SyntheticFrameProviderSP m_frame_provider_sp;
+  /// The Scripted Frame Providers for this thread.
+  std::vector<lldb::SyntheticFrameProviderSP> m_frame_providers;
```
Should we make this a `SmallVector<lldb::SyntheticFrameProviderSP, 0>`. Presumably most of the time we won't have any?

JDevlieghere commented on lldb/source/Target/Thread.cpp:
```diff
@@ -1505,29 +1505,39 @@ llvm::Error Thread::LoadScriptedFrameProvider(
+  //   provider's INPUT
+  StackFrameListSP input_frames;
+  if (m_frame_providers.empty()) {
+    // First provider gets real unwinder frames
```
Period

JDevlieghere commented on lldb/source/Target/Thread.cpp:
```diff
@@ -1505,29 +1505,39 @@ llvm::Error Thread::LoadScriptedFrameProvider(
 
-  ClearScriptedFrameProvider();
-  m_frame_provider_sp = *provider_or_err;
+  // Append to the chain
```
```suggestion
  // Append to the chain.
```

JDevlieghere reviewed (approved):
LGTM modulo comments

medismailben committed:
[lldb] Enable chaining multiple scripted frame providers per thread

This patch allows threads to have multiple SyntheticFrameProviderSP instances
that chain together sequentially. Each provider receives the output of the
previous provider as input, creating a transformation pipeline.

It changes `Thread::m_frame_provider_sp` to a vector, adds provider parameter
to SyntheticStackFrameList to avoid calling back into
`Thread::GetFrameProvider()` during frame fetching, updated
`LoadScriptedFrameProvider()` to chain providers by wrapping each previous
provider's output in a `SyntheticStackFrameList` for the next provider and
finally, loads ALL matching providers in priority order instead of just the
first one.

The chaining works as follows:
```
  Real Unwinder Frames
      ↓
  Provider 1 (priority 10) → adds/transforms frames
      ↓
  Provider 2 (priority 20) → transforms Provider 1's output
      ↓
  Provider 3 (priority 30) → transforms Provider 2's output
      ↓
  Final frame list shown to user
  ```

This patch also adds a test for this (test_chained_frame_providers) to verify
that 3 providers chain correctly: `AddFooFrameProvider`, `AddBarFrameProvider`,
`AddBazFrameProvider`.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

medismailben committed:
Merge branch 'main' into frame-provider-chaining

medismailben commented:
/cherry-pick 17b01bb cd70e2d 58f623c 8122d0e 10f2611 c373d76 943782be5aec5db854065145dd73a618c3a775be
